contrib/c-ares-cmake/CMakeLists.txt

```diff
@@ -1,35 +1,95 @@
+    "${LIBRARY_DIR}/src/lib/ares_parse_naptr_reply.c"
+    "${LIBRARY_DIR}/src/lib/ares_parse_ns_reply.c"
+    "${LIBRARY_DIR}/src/lib/ares_parse_ptr_reply.c"
+    "${LIBRARY_DIR}/src/lib/ares_parse_soa_reply.c"
+    "${LIBRARY_DIR}/src/lib/ares_parse_srv_reply.c"
+    "${LIBRARY_DIR}/src/lib/ares_parse_txt_reply.c"
+    "${LIBRARY_DIR}/src/lib/ares_parse_uri_reply.c"
+    "${LIBRARY_DIR}/src/lib/ares_platform.c"
+    "${LIBRARY_DIR}/src/lib/ares_process.c"
+    "${LIBRARY_DIR}/src/lib/ares_query.c"
+    "${LIBRARY_DIR}/src/lib/ares_search.c"
+    "${LIBRARY_DIR}/src/lib/ares_send.c"
+    "${LIBRARY_DIR}/src/lib/ares_strcasecmp.c"
+    "${LIBRARY_DIR}/src/lib/ares_strdup.c"
+    "${LIBRARY_DIR}/src/lib/ares_strerror.c"
+    "${LIBRARY_DIR}/src/lib/ares_strsplit.c"
+    "${LIBRARY_DIR}/src/lib/ares_timeout.c"
+    "${LIBRARY_DIR}/src/lib/ares_version.c"
+    "${LIBRARY_DIR}/src/lib/ares_writev.c"
+    "${LIBRARY_DIR}/src/lib/bitncmp.c"
+    "${LIBRARY_DIR}/src/lib/inet_net_pton.c"
+    "${LIBRARY_DIR}/src/lib/inet_ntop.c"
+    "${LIBRARY_DIR}/src/lib/windows_port.c"
+)
 
-# Disable looking for libnsl on a platforms that has gethostbyname in glibc
-#
-# c-ares searching for gethostbyname in the libnsl library, however in the
-# version that shipped with gRPC it doing it wrong [1], since it uses
-# CHECK_LIBRARY_EXISTS(), which will return TRUE even if the function exists in
-# another dependent library. The upstream already contains correct macro [2],
-# but it is not included in gRPC (even upstream gRPC, not the one that is
-# shipped with clickhousee).
-#
-#   [1]: https://github.com/c-ares/c-ares/blob/e982924acee7f7313b4baa4ee5ec000c5e373c30/CMakeLists.txt#L125
-#   [2]: https://github.com/c-ares/c-ares/blob/44fbc813685a1fa8aa3f27fcd7544faf612d376a/CMakeLists.txt#L146
-#
-# And because if you by some reason have libnsl [3] installed, clickhouse will
-# reject to start w/o it. While this is completelly different library.
-#
-#   [3]: https://packages.debian.org/bullseye/libnsl2
-if (NOT CMAKE_SYSTEM_NAME STREQUAL "SunOS")
-    set(HAVE_LIBNSL OFF CACHE BOOL "" FORCE)
+if (USE_STATIC_LIBRARIES)
+    add_library(_c-ares STATIC ${SRCS})
+    target_compile_definitions(_c-ares PUBLIC CARES_STATICLIB)
+else()
+    add_library(_c-ares SHARED ${SRCS})
+    target_compile_definitions(_c-ares PUBLIC CARES_BUILDING_LIBRARY)
 endif()
 
-# Force use of c-ares inet_net_pton instead of libresolv one
-set(HAVE_INET_NET_PTON OFF CACHE BOOL "" FORCE)
+target_compile_definitions(_c-ares PRIVATE HAVE_CONFIG_H=1)
 
-add_subdirectory("../c-ares/" "../c-ares/")
+target_include_directories(_c-ares SYSTEM PUBLIC
+    "${LIBRARY_DIR}/src/lib"
+    "${LIBRARY_DIR}/include"
+)
+
+# Platform-specific include directories. The original build system does a lot of checks to eventually generate two header files with defines:
+# ares_build.h and ares_config.h. To update, run the original CMake build in c-ares for each platform and copy the headers into the
```

Comment:
No nice way to call the configure from our CMake somehow?

Reply:
Before this PR, our CMake called the CMake of c-ares to create the c-ares library. The problem is that depending on the OS and the installed libs, c-ares's CMake produces different results.

The idea of "bundling" is to remove this variability as much as possible. So our CMake now builds c-ares directly but instead of probing the OS/packages/env and coming to different conclusions, the configuration is fixed and given by the pre-generated headers "ares_build.h" and "ares_config.h" which exist per-platform. The same technique is also used in other bundled 3rd party libraries.

Not sure if that answers your question :)

PS: One could argue that the whole "bundling" approach is a maintenance burden when 3rd party libs are upgraded but it is what it is. It helps at least with reproducible builds.

Reply:
Thanks for explaining, now it makes much more sense.